### PR TITLE
Optional expanded peep behaviour for walking and queuing

### PR DESCRIPTION
**Pull request by JeroenDStout**
I have created a branch in which I made some modifications to peep walking (not AI itself), which are exposed through (for now) five hidden settings in config.ini; they are off by default.

The idea of this alternate behaviour is to move peeps more off-grid, both for walking and queuing, and to add a little bit of extra character to them. Please note what I am showing here is not meant to be final. I would like to hear your thoughts on it!

messy time

Just to emphasize: _all_ of these options are written to have _no_ impact on path-finding or other higher AI. My main idea is to allow a player to _opt in_ to additional peep behaviour.

There are five changes:
- `peep_allow_sidestepping` (boolean)

Currently peeps walk straight lines and diagonals. This option allows them to 'side-step' by a pixel, meaning they have more freedom in walking lines. They also have a 'preferred' direction, meaning they are less inclined to alternate constantly between directions.
- `guest_messy_queuing` (boolean)

Currently peeps queue in perfect straight lines. This options allows them to queue messily, occasionally side by side. They try to keep a distance from the person ahead of them, which they prefer to be large; but if they are in a full, slow queue they succumb to pressure and start packing closely.

You can at a glance tell the difference between a ride with a high through-put and a ride which has more people queuing than places:

messy queuing fast
messy queuing busy

People (as is seen here) also skip around corners and queue on diagonals if the paths are quickly winding. Because of their dynamic preferences they constantly adjust the queue. In my experience this is what the rear of a queue experiences in a theme park; a constant slightly movement, whereas the head of the queue has great speed-ups and slow-downs.
- `peep_messy_walking` (boolean)

Peeps currently walk to the centre of a tile, but have a randomised threshold for getting there. This gives them some way of walking off-centre, but they frequently correct through (flickering) diagonals. The new option makes peeps walk the least possible to get to the new tile, meaning they always try to cut corners. They are also programmed to not take a turn instantly and instead incorporate a tiny soft turn.

messy walking

Especially when looking at a main square, they move less like cars in a grid-based city, and more like strollers in a park. When very nauseous they walk slightly less straight lines.
- `peep_messy_congestion` (boolean)

Currently peeps base walking pace only on their state. This option enables peeps who get congested (many peeps on one tile) to slow down. This makes avoiding tile-wide bottlenecks more interesting and shows bad design choices very quickly.
- `guest_max_time_in_park` (in minutes)

Currently guests stay in the park indefinitely. They tend to go for certain rides first, which means that with a lack of new peeps the most popular rides are left empty. Setting this to a lower value such as 300, after spending 6 hours in the park peeps decide they want to go home. This makes room for new blood and a more even distribution of peeps through the park.

This option calls `peep_leave_park`, which is also what happens if they leave the park through unhappiness.

---

These options are arguably best seen in-game. **_Make sure you enable them in config.ini._**

I have used `STOUT_PEEPS_EXPANDED_EXPERIMENT` as a macro to enable/disable the special code. This is not to put my name on it, rather it is a convenience that quickly shows no code is changed outside of the macros.

---

My own (humble) suggestion would be to allow for some expanded peep behaviour to be opt-in in the base game, perhaps even beyond what I have suggested above. Not everybody will want these options, but some interesting things are possible with rudimentary changes to peep behaviour: I believe quite a few people will greatly enjoy more messy queues, walking and other potential options.

Establishing such a thing as 'expanded peep behaviour' could be a good way to include these things without disrupting the base game.

I look forward to your reactions.

### CODE REVIEW

File: src/peep/peep.c [hunk 1]
```diff
@@ -5699,6 +5790,19 @@ static void peep_update(rct_peep *peep)
...
+		if (peep->peeps_ex_crowded_store > 15)
+			stepsToTake /= 3 + scenario_rand_max(5);
+		else if (peep->peeps_ex_crowded_store > 9)
+			stepsToTake /= 2 + scenario_rand_max(3);
+		else if (peep->peeps_ex_crowded_store > 4) {
+			stepsToTake *= 2;
+			stepsToTake /= 3 + scenario_rand_max(2);
+		}
+	}
+	#endif STOUT_PEEPS_EXPANDED_EXPERIMENT
```
Comment (IntelOrca) [inline src/peep/peep.c]: This is what's causing Travis to fail. I am surprised MSVC is ok with this.

**Comment by IntelOrca** [comment]
Currently we have no way of synchronising configuration options easily from server to client in multiplayer other than a hack which pushes extra data on the map stream. We might want to come up with something better than that first.

**Comment by IntelOrca** [comment]
Chrome and other Google products have a kind of labs or experiments page (e.g. `chrome://flags/`). Perhaps we should do something similar.

File: src/peep/peep.c (hunk 1 above)
Comment (JeroenDStout) [inline src/peep/peep.c]: I must admit I cannot see the problem. Not to say there isn't one, I just cannot see it :-\

**Comment by JeroenDStout** [comment]
> Chrome and other Google products have a kind of labs or experiments page (e.g. chrome://flags/). Perhaps we should do something similar.

That was sort-of the idea of the hidden config settings, for now. Especially seeing it as a 'expanded' feature set, so optional.

Although I can definitely see the virtue in waiting for a 'exchange settings' feature. Perhaps if that is the long-term solution, giving a fat warning if you go online with 'lab' things enabled might be the interim?

File: src/peep/peep.c (hunk 1 above)
Comment (JeroenDStout) [inline src/peep/peep.c]: Right, the `STOUT_PEEPS_EXPANDED_EXPERIMENT`. I suppose MSVC thinks that is defined to nothing, so it also doesn't care that it is there.

**Comment by CraigCraig** [comment]
I just wanted to note, I have tested this for hours today, I have had no issues so far :)

File: src/peep/peep.c [hunk 2]
```diff
@@ -1000,22 +1017,84 @@ int peep_update_action(sint16* x, sint16* y, sint16* xy_distance, rct_peep* peep
...
-		int direction = 0;
-		if (x_delta < y_delta){
-			direction = 8;
-			if (*y >= 0){
-				direction = 24;
-			}
+
+	#ifdef STOUT_PEEPS_EXPANDED_EXPERIMENT
+		if (!gConfigPeepsEx.peep_allow_sidestepping) {
+	#endif
```
Comment (duncanspumpkin) [inline src/peep/peep.c]: dont do braces like this across #ifdefs its super confusing to read.

**Comment by Phyxion** [comment]
This looks great and makes the game much more alive. I like it :+1:

File: src/peep/peep.c (hunk 2 above)
Comment (JeroenDStout) [inline src/peep/peep.c]: You are right, I will find a more elegant way of doing that.

**Comment by jensj12** [comment]
The distance between guests in the queue doesn't feel realistic to me.
People don't keep moving slowly towards the next person in queue but instead move forward every now and then, moving closer to the ones in front of them if they haven't moved up yet. Especially when I'm talking with someone I just move forward at random times, not knowing what the people in front of me did.
People near the end of the queue don't look at their watch (and miss the train like near the giga-coaster) but get super exited and push forward as far as possible. Stations are the busiest parts of the queue, even for high-capacity rides.
For the normal paths, the flickering is indeed a bit too much, but other than that, great work! Perhaps make them always walk straight forward if they don't change direction?

**Comment by JeroenDStout** [comment]
@Phyxion 

I thought of a 'expanded behaviour' tab in the options menu, but I think (as @IntelOrca suggested) a 'labs' page could make room for many options, perhaps?

@jensj12 

> People near the end of the queue don't look at their watch (and miss the train like near the giga-coaster) but get super exited and push forward as far as possible. Stations are the busiest parts of the queue, even for high-capacity rides.

That's a very good point. I have thought of forcing compression towards the head of the queue, but I didn't get to implementing it.

Unfortunately the checking the watch is in another part of the code, and it happens in vanilla. Although I guess it is possible to stop that from happening, it is just harder.

> People don't keep moving slowly towards the next person in queue but instead move forward every now and then, moving closer to the ones in front of them if they haven't moved up yet. Especially when I'm talking with someone I just move forward at random times, not knowing what the people in front of me did.

Maybe a change that could be made is to make it less likely peeps reduce queuing distance while waiting, but increase the amount they reduce it by. I agree the shuffle-stepping is sort-of off.

Here's the queue with some adjustments. They now compress at the head of the queue if they think they might get on the next train:

messy queuing fast 2

With these tweeks they rely more on distance propagation (people at the back breathing down your neck) and low-chance (large) forward stepping; + compression when very close to the head.

I would love a variable to be set for each station on a ride that tells whether there is a car, so people can only compress when they are super-close to being on-board.

> For the normal paths, the flickering is indeed a bit too much, but other than that, great work! Perhaps make them always walk straight forward if they don't change direction?

It turns out the flicking was a bug that survived from the main game! The problem was they have a distance check to their target, but sometimes (as in vanilla) they actually walk past it so they flicker back.

It should be solved now:

messy walking 2

I did give them a favoured direction, so if they are already walking in direction N, they will greatly favour continuing in direction N.

These are really good comments! If you have any more thoughts (perhaps also in-game) I would love to hear them. This will be WIP for a while, I think.

**Comment by Phyxion** [comment]
Maybe they should favor the right side of the path (and some still ignore that)?

**Comment by marijnvdwerf** [comment]
@Phyxion And favour the left for specific currencies? ;)

**Comment by JeroenDStout** [comment]
If it is quite busy now they do favour one side. It is not very noticeable. I could make them hold path sides based on their sprite ID (so they do it consistently, but randomly).

**Comment by JeroenDStout** [comment]
Accidentally closed this one through a mis-click (apparently!).

**Comment by Phyxion** [comment]
@marijnvdwerf Yes, UK should be messy, Australian should be left, rest of world is probably mostly right :)

**Comment by JeroenDStout** [comment]
I think I might prefer to put the path-side options in the config, rather than read the monetary settings or such. :)

**Comment by jensj12** [comment]
Different settings in a multiplayer game might cause a desync. The easiest solution would be to not make it a setting.

**Comment by CraigCraig** [comment]
Perhaps have a check for certain settings when joining multiplayer to toggle them off? A tool-tip warning on the setting saying: "Not Multiplayer-Compatible"

**Comment by JeroenDStout** [comment]
> Different settings in a multiplayer game might cause a desync. The easiest solution would be to not make it a setting.

> Perhaps have a check for certain settings when joining multiplayer to toggle them off? Also perhaps a tool-tip warning on the setting saying: "Not Multiplayer-Compatible"

I was thinking part of a server's announcement could be its usage of extension (lab) settings. Clicking on the server could require the client choose to adjust his settings to the server's before attempting to join. But perhaps I could make a separate issue for this.

**Comment by IntelOrca** [comment]
@JeroenDStout In OpenTTD, joining a sever will automatically change some of the advance settings to the server's settings and lock them from being changed. It reverts back when the player leaves.

**Comment by janisozaur** [comment]
I haven't at the code yet, but an important thing to note: our networking code depends on simulation going on *exactly the same* across all connected clients, otherwise it would cause a "desync", where clients have different logic states.

If your code does not affect logic, as in: only the way peeps look and they act and pathfind like before, then this is OK, but otherwise this will need a thorough review.

Another note on that is, if this only affect peeps display, then there's nothing really to sync over network and clients can enable it on their own.

Keep in mind this is unacceptable to require client to restart their game with some option passed on startup, so if this has to be an option, you'll need to make it possible to change it on the fly (doesn't have to be mid-game, on scenario load is just fine, see https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/rct2/S6Importer.cpp#L471)

**Comment by JeroenDStout** [comment]
> In OpenTTD, joining a sever will automatically change some of the advance settings to the server's settings and lock them from being changed. It reverts back when the player leaves.

That sounds like a very good thing to me. Cultivating an expanded behaviour set may be a nice addition.

@janisozaur 

> If your code does not affect logic, as in: only the way peeps look and they act and pathfind like before, then this is OK, but otherwise this will need a thorough review.

It does certainly affect behaviour, the pathing inside tiles is different. I believe I only call deterministic functions (scenario-rand, scenario ticks, etc), but as I am very new to the engine I wouldn't take my word for it.

> Keep in mind this is unacceptable to require client to restart their game with some option passed on startup, so if this has to be an option, you'll need to make it possible to change it on the fly

Right now whether or not logic is applied is taken directly from the settings, so it is possible to change them on the fly. They could be exposed in the interface right now, but I haven't done that as I was not sure whether a labs page or 'expanded behaviour' page was desirable.

**Comment by Joshwoo70** [comment]
could we add a epic tag and some WIP tags for this? thanks.

**Comment by AaronVanGeffen** [comment]
I think the changes are too subtle for an epic tag, but the issue should probably be tagged, yes. I don't believe it's WIP any longer, but pending review, though?

**Comment by duncanspumpkin** [comment]
It still requires a large rethink on how we should do this. We don't want to clutter the codebase with everyone's alterations to the ai.

**Comment by Gymnasiast** [comment]
Unfortunately, we have to close this. Our Peep AI still needs quite a bit of work to improve it, we're not nearly at a point where we can introduce jitter like this. The PR has also gone stale.

Eventually, we might have a way to switch AI systems like OpenTTD does.

**Pull request closed without merging**